psi4/src/psi4/lib3index/df_helper.cc

```diff
@@ -1551,282 +1747,303 @@ void DF_Helper::transform_core() {
+                            C_DGEMM('N', 'N', a1, a2, a1, 1.0, metp, naux, &Np[i*a1*a2], a2, 
+                                0.0, &Lp[i*a1*a2], a2);
                         }
                     }
-
-                    // (Qm)(mb)->(Qb)
-                    C_DGEMM('N', 'N', block_size, bsize, sp_size, 1.0, &Mp[jump], sp_size, &C_buffers[rank][0], bsize,
-                            0.0, &Tp[k * block_size * bsize], bsize);
                 }
-                timer_off("DFH: First Contraction");
-                timer_off("DFH: Total Transform");
-                timer_off("DFH: Total Workflow");
-
-                // to completion per transformation
-                for (size_t k = 0; k < strides_[i]; k++) {
-                    std::string left = std::get<0>(transf_[order_[count + k]]);
-                    std::string right = std::get<1>(transf_[order_[count + k]]);
-                    std::tuple<SharedMatrix, size_t> I = (bspace.compare(left) == 0 ? spaces_[right] : spaces_[left]);
-
-                    size_t wsize = std::get<1>(I);
-                    double* Wp = std::get<0>(I)->pointer()[0];
+            }
+        }
+    }
+    timer_off("DFH: transform()");
+    transformed_ = true;
+}
 
-                    // (wp)(p|Qb)->(w|Qb)
-                    timer_on("DFH: Total Workflow");
-                    timer_on("DFH: Total Transform");
-                    timer_on("DFH: Second Contraction");
-                    C_DGEMM('T', 'N', wsize, block_size * bsize, nao_, 1.0, Wp, wsize, Tp, block_size * bsize, 0.0, Fp,
-                            block_size * bsize);
-                    timer_off("DFH: Second Contraction");
-                    timer_off("DFH: Total Transform");
-                    timer_off("DFH: Total Workflow");
+void DF_Helper::first_transform_pQq(int nao, int naux, int bsize, int bcount, int block_size, int rank, 
+    double* Mp, double* Tp, double* Bp, std::vector<std::vector<double>> C_buffers){
+    
+    // perform first contraction on pQq, thread over p.
+    #pragma omp parallel for firstprivate(nao, naux, bsize, \
+                          block_size) private(rank) schedule(guided) num_threads(nthreads_)
+    for (size_t k = 0; k < nao_; k++) {
+        
+        // truncate transformation matrix according to fun_mask
+        size_t sp_size = small_skips_[k];
+        size_t jump = (AO_core_ ? big_skips_[k] + bcount * sp_size : (big_skips_[k] * block_size) / naux_);
 
-                    // setup putf
-                    std::string putf =
-                        (!direct_ ? std::get<1>(files_[order_[count + k]]) : std::get<0>(files_[order_[count + k]]));
+        #ifdef _OPENMP
+        rank = omp_get_thread_num();
+        #endif
+        for (size_t m = 0, sp_count = -1; m < nao_; m++) {
+            if (schwarz_fun_mask_[k * nao_ + m]) {
+                sp_count++;
+                C_DCOPY(bsize, &Bp[m * bsize], 1, &C_buffers[rank][sp_count * bsize], 1);
```

Comment:
If we run this through VTune, how much time do we spending copying these orbitals to the new format? This is an aligned copy, so much better than previous implementations with strides copies. However, any rank 1 operation these days is pretty painful.

Reply:
Actually, this copy also occurs in previous implementations:
https://github.com/psi4/psi4/blob/4c3ce1aa1fb8fe6914448071af33d1ed7cf7c5f2/psi4/src/psi4/libfock/DFJK.cc#L1823
`DF_Helper` avoids the strided copy altogether.  
Still, profiling this on VTune is necessary, will get back to you about that.